File: lang/stlc/src/traits/free_vars/list.rs
use super::FreeVars;
use crate::{
    syntax::{Cons, Head, IsNil, Nil, Tail},
    Var,
};
use std::collections::HashSet;

impl FreeVars for Nil {
    fn free_vars(&self) -> HashSet<Var> {
        HashSet::new()
    }
}

impl FreeVars for Cons {
    fn free_vars(&self) -> HashSet<Var> {
        let mut vars = self.fst.free_vars();
        vars.extend(self.rst.free_vars());
        vars
    }
}

impl FreeVars for IsNil {
    fn free_vars(&self) -> HashSet<Var> {
        self.list.free_vars()
    }
}

impl FreeVars for Head {
    fn free_vars(&self) -> HashSet<Var> {
        self.list.free_vars()
    }
}

impl FreeVars for Tail {
    fn free_vars(&self) -> HashSet<Var> {
        self.list.free_vars()
    }
}

File: examples/featherweight/list/list.java
class List extends Object{
  List() { super(); } 

  int head() { return 0; }

  List tail() { return new List() };
}

class Nil extends List { 
  Nil() { super(); }
}

class Cons extends List{
  int first;
  List rest;

  Cons(int first,List rest) {
    super(); 
    self.first = first;
    self.rest= rest;
  }

  int head() {
    return self.first
  }

  List tail(){
    return sefl.rest
  }
}
